README.md

```diff
@@ -140,6 +123,35 @@ In order to facilitate easier adoption, script hashes should also be added to CS
 This necessitates adding two fields: the hash of the content of a script (for inline/eval scripts), and the hash of the URL (initial URL, without following redirects) of the script (for script src elements).
 
 
+### Backwards compatibility
+One of the goals is for the new features to be deployable on sites without breaking on older browsers. There are two ways to achieve this:
+
+
+#### Extend script-src with new features, implemented in a way that old browsers can still see a more lax policy
+This would extend the existing script-src directive with 2 new keywords, `eval-<hashalgorithm>-<hash>` and `url-<hashalgorithm>-<hash>`. For eval, backwards compatibility would be achieved by browsers that support the new features ignore `unsafe-eval` if one or more eval hashes are present (which matches the existing behavior for `unsafe-inline`, which is [ignored if script hashes are present](https://www.w3.org/TR/CSP3/#allow-all-inline)). This would allow a site that needs to use eval to set both unsafe-eval and the script hashes. Browsers that don't support the new feature would still work due to the presence of unsafe-eval (and ignore the hashes since they don't understand that keyword), while browsers that support hashes would ignore `unsafe-eval`. In practice this would look like:
+
+
+```
+Content-Security-Policy: script-src 'eval-sha256-SHA256("console.log('Hello World')';
+<script>
+eval("console.log('Hello World')");
+</script>
+```
+
+
+For URL hashes, a similar approach can be followed. Browsers that support the new features can ignore host based allowlisiting if a URL hash is present. Unlike eval though, this leaves some open questions about interactions with other directives, notably `strict-dynamic`. A common fallback for sites that use URL hashes would be to include `https:` in the fallback policy. However sites might require `strict-dynamic` in order to use URL hashes, and they can't set a policy like `script-src https: strict-dynamic, url-sha256-<hash>` since in older browsers `strict-dynamic` would cause `https:` to be [ignored](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/script-src#strict-dynamic). This can be solved by either:
+
+- Adding a new keyword that enables strict-dynamic like behavior for URL hashes, but is ignored in older browsers, e.g. `strict-dynamic-hashes`.
+- Having the presence of URL hashes imply strict-dynamic, so the keyword is not necessary. This would be the easiest behavior to explain, but it means that sites using hashes lose the option to disable `strict-dynamic` behavior, so sites that don't need it, but want to use URL hashes have a policy that is more lax than they really need.
```

Comment:
A variant of this is for url-hashes to imply strict-dynamic but only for scripts loaded by url hashes.

So, something like:

`script-src url-sha256(example.com/script.js) nonce-abc`

Would apply strict-dynamic like behavior to example.com/script.js, but not to scripts with nonce=abc.

(This might already be what you intended, in which case it might be worth stating this explicity)

Reply:
Done